rfcs/0030-fork-signalling.md

```diff
@@ -0,0 +1,94 @@
+## Summary
+[summary]: #summary
+
+Nodes can signal to other nodes that they intend to switch from the current fork to a new fork.
+
+## Motivation
+[motivation]: #motivation
+
+Each node in the network has a notion of the current fork it's running on. A node may wish to
+change forks, perhaps because it's running new software, or to create a new subnet of nodes with
+some common interest. Such an ability is a limited form of on-chain governance. When a node
+proposes a new fork, other nodes need to be informed of the proposal.
+
+## Detailed design
+[detailed-design]: #detailed-design
+
+A fork is denoted by a fork identifier (`fork ID`). Blocks have two fork ID fields, one to
+indicate the current fork, another to indicate a proposed fork. Because there will usually
+not be a proposed fork, that field has an option type.
+
+When a node is started, the current fork ID is retrieved from the node's configuration.
+If the node is being run for the first time, the fork ID must be provided as a command-line
+flag. Alternatively, a current fork ID can be provided via a compile-time constant,
+which can be overridden via the configuration or command-line flag.
+
+For RPCs between nodes, if the response contains a block with a different current fork ID,
+that response is ignored, and the sender punished. The next fork ID is ignored for these RPCs.
+
+For gossiped blocks, if the block's current fork ID differs from the node's current fork ID,
+that block is ignored, and the sender is punished.
+
+To signal a proposed fork change, there needs to be a mechanism for the node to set its
+next fork ID, to be included in the blocks it gossips. That can be accomplished by a
+client command that stores the next fork ID in the configuration, to be used the
+next time the node is run. Additionally, we may wish to change a node's next fork ID
+dynamically via a GraphQL query, which would also store the next fork ID in the configuration.
+
+A node can track the number of gossiped blocks with a proposed next fork ID and elect to
+restart with a new current fork ID. Nodes should log statistics of proposed next fork IDs that
+appear in received gossip, so that node operators can make informed choices whether
+to switch. Such statistics should also be available via GraphQL. We defer discussion
+of the details of such statistics.
+
+Some of this design appears in PR #4347, which has not been merged as of this writing.
+The punishment implemented there for mismatched current fork IDs is an Instaban, which
```

Comment:
I agree, and in hindsight, I do not think this should be an instaban. See my response in the `Unresolved Questions` section.

Reply:
Changed to trust decrease of 0.25 in #4347.